persistence/src/main/java/com/epam/esm/persistence/repository/impl/TagRepositoryImpl.java

```diff
@@ -1,88 +1,98 @@
 package com.epam.esm.persistence.repository.impl;
 
-import com.epam.esm.domain.GiftCertificate;
-import com.epam.esm.domain.Tag;
-import com.epam.esm.persistence.repository.ColumnName;
-import com.epam.esm.persistence.repository.RepositoryException;
+import com.epam.esm.model.GiftCertificate;
+import com.epam.esm.model.Tag;
 import com.epam.esm.persistence.repository.TagRepository;
-import org.springframework.beans.factory.annotation.Autowired;
-import org.springframework.beans.factory.annotation.Qualifier;
-import org.springframework.jdbc.core.JdbcTemplate;
-import org.springframework.jdbc.core.RowMapper;
-import org.springframework.jdbc.core.simple.SimpleJdbcInsert;
 import org.springframework.stereotype.Repository;
 
-import java.util.HashMap;
+import javax.persistence.EntityManager;
+import javax.persistence.PersistenceContext;
+import javax.persistence.Query;
+import javax.persistence.TypedQuery;
+import java.util.ArrayList;
 import java.util.List;
-import java.util.Map;
 
 @Repository
 public class TagRepositoryImpl implements TagRepository {
-    private static final String SELECT_ALL_QUERY = "SELECT * FROM Tag";
-    private static final String SELECT_BY_ID_QUERY = "SELECT * FROM Tag WHERE id=?";
-    private static final String SELECT_BY_NAME_QUERY = "SELECT * FROM Tag WHERE name=?";
-    private static final String DELETE_QUERY = "DELETE FROM Tag WHERE id=?";
-    private static final String SELECT_ASSOCIATED_CERTIFICATES =
-            "SELECT * FROM Gift_certificate\n" +
-                    "JOIN Gift_certificate_has_Tag GchT on Gift_certificate.id = GchT.certificate\n" +
-                    "WHERE GchT.tag=?";
-    private static final int MIN_AFFECTED_ROWS = 1;
+    private static final String SELECT_ALL_QUERY = "SELECT tag FROM Tag tag";
+    private static final String SELECT_BY_NAME_QUERY = "SELECT tag FROM Tag tag WHERE tag.name=:tagName";
+    private static final String COUNT_ALL_QUERY = "SELECT COUNT(tag) FROM Tag tag";
 
-    private final JdbcTemplate jdbcTemplate;
-    private final RowMapper<Tag> tagRowMapper;
-    private final RowMapper<GiftCertificate> certificateRowMapper;
-    private final SimpleJdbcInsert jdbcInsert;
+    /**
+     * Query to select most widely used tag(s) of a user with the highest cost of all orders.
+     */
+    private static final String MOST_POPULAR_QUERY = "SELECT Tag.id, Tag.name FROM Tag\n" +
+            "JOIN Gift_certificate_has_Tag GchT on Tag.id = GchT.tag AND GchT.certificate IN (SELECT Gift_certificate" +
+            ".id FROM Gift_certificate\n" +
+            "JOIN Orders O on Gift_certificate.id = O.certificate_id AND O.user_id = (\n" +
+            "        SELECT User.id FROM User\n" +
+            "        JOIN Orders O on User.id = O.user_id\n" +
+            "        JOIN Gift_certificate Gc on Gc.id = O.certificate_id\n" +
+            "        GROUP BY User.id\n" +
+            "        HAVING SUM(O.total_price)\n" +
```

Comment:
What purpose of having is here?

Reply:
There is no sense in `HAVING` statement here.